src/bpfbytecode.cpp

```diff
@@ -26,6 +45,93 @@ const std::vector<uint8_t> &BpfBytecode::getSection(
   return sections_.at(name);
 }
 
+int BpfBytecode::create_maps()
+{
+  int failed_maps = 0;
+  for (auto &map : maps_) {
+    LIBBPF_OPTS(bpf_map_create_opts, opts);
+    map.second.fd = bpf_map_create(static_cast<::bpf_map_type>(
+                                       map.second.type()),
+                                   map.second.bpf_name().c_str(),
+                                   map.second.key_size(),
+                                   map.second.value_size(),
+                                   map.second.max_entries(),
+                                   &opts);
+    if (map.second.fd < 0) {
+      failed_maps++;
+      LOG(ERROR) << "failed to create map: '" << map.first
+                 << "': " << strerror(errno);
+    }
+  }
+
+  if (failed_maps > 0) {
+    std::cerr << "Creation of the required BPF maps has failed." << std::endl;
+    std::cerr << "Make sure you have all the required permissions and are not";
+    std::cerr << " confined (e.g. like" << std::endl;
+    std::cerr << "snapcraft does). `dmesg` will likely have useful output for";
+    std::cerr << " further troubleshooting" << std::endl;
+  }
+
+  return failed_maps;
+}
+
+bool BpfBytecode::hasMap(MapType internal_type) const
+{
+  return maps_.find(to_string(internal_type)) != maps_.end();
+}
+
+bool BpfBytecode::hasMap(const StackType &stack_type) const
+{
+  return maps_.find(stack_type.name()) != maps_.end();
+}
+
+const BpfMap &BpfBytecode::getMap(const std::string &name) const
+{
+  auto map = maps_.find(name);
+  if (map == maps_.end()) {
+    throw std::runtime_error("Unknown map: " + name);
+  }
+  return map->second;
+}
+
+const BpfMap &BpfBytecode::getMap(MapType internal_type) const
+{
+  return getMap(to_string(internal_type));
+}
+
+const BpfMap &BpfBytecode::getMap(int map_id) const
+{
+  auto map = maps_by_id_.find(map_id);
+  if (map == maps_by_id_.end()) {
+    throw std::runtime_error("Unknown map id: " + std::to_string(map_id));
+  }
+  return *map->second;
+}
+
+const std::map<std::string, BpfMap> &BpfBytecode::maps() const
+{
+  return maps_;
+}
+
+int BpfBytecode::countStackMaps() const
+{
+  int n = 0;
+  for (auto &map : maps_) {
+    if (map.second.is_stack_map())
+      n++;
+  }
+  return n;
+}
+
+void BpfBytecode::set_map_ids(RequiredResources &resources)
+{
+  for (auto &map : maps_) {
+    auto map_info = resources.maps_info.find(map.first);
+    if (map_info != resources.maps_info.end() && map_info->second.id != -1)
+      maps_by_id_.emplace(map_info->second.id, &map.second);
```

Comment:
IIUC, this should be fine b/c references to `std::map` are not invalidated by inserting new elements. But @ajor is the C++ expert here so I'd like to highlight this so that you double check.

Reply:
Yep, looks fine to me 👍